core/controllers/reader.py

```diff
@@ -274,15 +274,51 @@ def get(self, exploration_id):
         exploration_rights = rights_manager.get_exploration_rights(
             exploration_id, strict=False)
         user_settings = user_services.get_user_settings(self.user_id)
+        exp_user_data = exp_fetchers.get_exploration_user_data(
+            self.user_id, exploration_id)
 
         preferred_audio_language_code = None
         preferred_language_codes = None
+        has_viewed_lesson_info_modal_once = None
 
         if user_settings is not None:
             preferred_audio_language_code = (
                 user_settings.preferred_audio_language_code)
             preferred_language_codes = (
                 user_settings.preferred_language_codes)
+            has_viewed_lesson_info_modal_once = (
+                user_settings.has_viewed_lesson_info_modal_once)
+
+        furthest_reached_checkpoint_exp_version = None
+        furthest_reached_checkpoint_state_name = None
+        most_recently_reached_checkpoint_exp_version = None
+        most_recently_reached_checkpoint_state_name = None
+
+        # If exp_user_data is None, it means the exploration is started
+        # for the first time and no checkpoint progress of the user exists for
+        # the respective exploration.
+        # Exploration version 'v' passed as a parameter in GET request means a
+        # previous version of the exploration is being fetched. In that case,
+        # we want that exploration to start from the beginning and do not
+        # allow users to save their checkpoint progress in older exploration
+        # version.
+        if exp_user_data is not None and version is None:
+            # If the latest exploration version is ahead of the most recently
+            # interacted exploration version.
+            if (exp_user_data.most_recently_reached_checkpoint_exp_version is not None and ( # pylint: disable=line-too-long
+                    exp_user_data.most_recently_reached_checkpoint_exp_version < exploration.version)): # pylint: disable=line-too-long
+                exp_user_data = (
```

Comment:
Do not override the variable 
```suggestion
                updated_exp_user_data = (
```

Reply:
Used another variable `synced_exp_user_data` for the same now. PTAL